pex/resolve/pep_691/api.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import json
+
+from pex.compatibility import HTTPError, text, urlparse
+from pex.fetcher import URLFetcher
+from pex.pep_440 import Version
+from pex.pep_503 import ProjectName
+from pex.resolve.pep_691.model import Endpoint, File, Meta, Project
+from pex.resolve.resolved_requirement import Fingerprint
+from pex.sorted_tuple import SortedTuple
+from pex.third_party.packaging.version import Version as PackagingVersion
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING, cast
+
+if TYPE_CHECKING:
+    from typing import Any, Dict, Optional, Type, TypeVar
+
+    import attr  # vendor:skip
+
+    _V = TypeVar("_V")
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class Client(object):
+    class Error(Exception):
+        """Indicates an error accessing or parsing the results of a PEP-691 API endpoint."""
+
+    ACCEPT = ("application/vnd.pypi.simple.v1+json",)
+
+    _url_fetcher = attr.ib(factory=URLFetcher)  # type: URLFetcher
+
+    def request(self, endpoint):
+        # type: (Endpoint) -> Project
+        """Fetches PEP-691 project file metadata from the endpoint.
+
+        Raises :class:`Client.Error` if there is a problem accessing the endpoint or parsing the
+        results returned from it.
+        """
+
+        if endpoint.content_type not in self.ACCEPT:
+            raise ValueError(
+                "Asked to request project metadata from {url} for {content_type} but only the "
+                "following API content types are accepted: {accepted}".format(
+                    url=endpoint.url,
+                    content_type=endpoint.content_type,
+                    accepted=", ".join(self.ACCEPT),
+                )
+            )
+
+        def request_error(msg):
+            # type: (str) -> Client.Error
+            return self.Error(
+                "PEP-691 API request to {url} for {content_type} {msg}".format(
+                    url=endpoint.url, content_type=endpoint.content_type, msg=msg
+                )
+            )
+
+        try:
+            with TRACER.timed(
+                "Fetching PEP-691 index metadata from {url} for {content_type}".format(
+                    url=endpoint.url, content_type=endpoint.content_type
+                )
+            ):
+                with self._url_fetcher.get_body_stream(
+                    endpoint.url, extra_headers={"Accept": endpoint.content_type}
```

Comment:
FWIW these requests average ~200ms on my CenturyLink mildly crappy home DSL. I get a parallelism of 10 and so for pantsbuild.pants with no cache I see ~1.2s to fetch all the fingerprints. Cached is sub-ms.